sentry-android-replay/src/main/java/io/sentry/android/replay/ReplayIntegration.kt

```diff
@@ -0,0 +1,317 @@
+package io.sentry.android.replay
+
+import android.content.Context
+import android.graphics.Bitmap
+import android.graphics.Point
+import android.graphics.Rect
+import android.os.Build
+import android.os.Build.VERSION
+import android.os.Build.VERSION_CODES
+import android.view.WindowManager
+import io.sentry.DateUtils
+import io.sentry.Hint
+import io.sentry.IHub
+import io.sentry.Integration
+import io.sentry.ReplayRecording
+import io.sentry.SentryLevel.DEBUG
+import io.sentry.SentryLevel.INFO
+import io.sentry.SentryOptions
+import io.sentry.SentryReplayEvent
+import io.sentry.protocol.SentryId
+import io.sentry.rrweb.RRWebMetaEvent
+import io.sentry.rrweb.RRWebVideoEvent
+import io.sentry.transport.ICurrentDateProvider
+import io.sentry.util.FileUtils
+import java.io.Closeable
+import java.io.File
+import java.util.Date
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.ThreadFactory
+import java.util.concurrent.TimeUnit.MILLISECONDS
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.atomic.AtomicReference
+import kotlin.LazyThreadSafetyMode.NONE
+import kotlin.math.roundToInt
+
+class ReplayIntegration(
+    private val context: Context,
+    private val dateProvider: ICurrentDateProvider
+) : Integration, Closeable, ScreenshotRecorderCallback {
+
+    companion object {
+        const val VIDEO_SEGMENT_DURATION = 5_000L
+        const val VIDEO_BUFFER_DURATION = 30_000L
+    }
+
+    private lateinit var options: SentryOptions
+    private var hub: IHub? = null
+    private var recorder: WindowRecorder? = null
+    private var cache: ReplayCache? = null
+
+    // TODO: probably not everything has to be thread-safe here
+    private val isEnabled = AtomicBoolean(false)
+    private val isRecording = AtomicBoolean(false)
+    private val currentReplayId = AtomicReference<SentryId>()
+    private val segmentTimestamp = AtomicReference<Date>()
+    private val currentSegment = AtomicInteger(0)
+    private val saver =
+        Executors.newSingleThreadScheduledExecutor(ReplayExecutorServiceThreadFactory())
+
+    private val screenBounds by lazy(NONE) {
+        // PixelCopy takes screenshots including system bars, so we have to get the real size here
+        val wm = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
+        if (VERSION.SDK_INT >= VERSION_CODES.R) {
+            wm.currentWindowMetrics.bounds
+        } else {
+            val screenBounds = Point()
+            @Suppress("DEPRECATION")
+            wm.defaultDisplay.getRealSize(screenBounds)
+            Rect(0, 0, screenBounds.x, screenBounds.y)
+        }
+    }
+
+    private val aspectRatio by lazy(NONE) {
+        screenBounds.bottom.toFloat() / screenBounds.right.toFloat()
```

Comment:
nit, would be semantically more correct:
```suggestion
        screenBounds.height.toFloat() / screenBounds.width.toFloat()
```

Reply:
nice, I thought `Rect` doesn't have those 👍